### PR TITLE
Fix deprecation logging of password policy.

### PR DESCRIPTION
**Pull request by mashhurs**
## Release notes

[rn:skip]

## What does this PR do?

Fixes the deprecation logging of the password policy. The issue is, when we upgrade Logstash (eg. 8.3.0 to 8.3.1+) current deprecation logging will be produced even default value (`WARN`) of `api.auth.basic.password_policy.mode` changes to `ERROR`.

## Why is it important/What is the impact to the user?

Does not produce the (wrong log) deprecation noise to end users.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
~~- [ ] I have made corresponding changes to the documentation~~
~- [ ] I have made corresponding change to the default configuration files (and/or docker env variables)~
- [x] I have added tests that prove my fix is effective or that my feature works

## Author's Checklist

- [ ]

## How to test this PR locally

- Enable the basic auth type and provide credentials: `api.auth.type: basic`
- Run Logstash and find the deprecation log, outcome should look like:
```
[2022-05-24T15:52:39,996][WARN ][deprecation.logstash.webserver] The default value of `api.auth.basic.password_policy.mode` will change to `ERROR` in a future release of Logstash. Set it to `ERROR` to enforce future requirement now.
```

## Related issues

- #13884

### CODE REVIEW

File: logstash-core/lib/logstash/webserver.rb [hunk 1]
```diff
@@ -105,6 +106,14 @@ def self.required_setting(settings, setting_name, trigger)
       settings.get(setting_name) || fail(ArgumentError, "Setting `#{setting_name}` is required when `#{trigger}` is set to `#{settings.get(trigger)}`. Please provide it in your `logstash.yml`")
     end
 
+    def self.required_setting_with_changing_default(settings, name, trigger, future_value)
+      effective_value = required_setting(settings, name, trigger)
+      if !settings.set?(name)
+        deprecation_logger.deprecated("The default value of `#{name}` will change to `#{future_value}` in a future release of Logstash. Set it to `#{future_value}` to enforce future requirement now.")
```
Comment (yaauie) [inline logstash-core/lib/logstash/webserver.rb:112]: A deprecation log should not only inform a user of an upcoming breaking change, it should empower them to address it by providing a clear path to either consuming the breaking change early or avoiding the breaking change by explicitly locking in the current behavior.
```suggestion
        deprecation_logger.deprecated("The default value of `#{name}` will change to `#{future_value}` in a future release of Logstash. Set it to `#{future_value}` to observe the future behavior early, or set it to `#{effective_value}` to lock in the current behavior.")
```

File: logstash-core/lib/logstash/webserver.rb [hunk 2]
```diff
@@ -105,6 +106,14 @@ def self.required_setting(settings, setting_name, trigger)
       settings.get(setting_name) || fail(ArgumentError, "Setting `#{setting_name}` is required when `#{trigger}` is set to `#{settings.get(trigger)}`. Please provide it in your `logstash.yml`")
     end
 
+    def self.required_setting_with_changing_default(settings, name, trigger, future_value)
+      effective_value = required_setting(settings, name, trigger)
+      if !settings.set?(name)
+        deprecation_logger.deprecated("The default value of `#{name}` will change to `#{future_value}` in a future release of Logstash. Set it to `#{future_value}` to observe the future behavior early, or set it to `#{effective_value}` to lock in the current behavior.")
```
Comment (yaauie) [inline logstash-core/lib/logstash/webserver.rb:112]: @karenzone can I get your review on the wording for this deprecation log?

As called it is effectively:

~~~
[2022-05-24T15:52:39,996][WARN ][deprecation.logstash.webserver] The default value of `api.auth.basic.password_policy.mode` will change to `ERROR` in a future release of Logstash. Set it to `ERROR` to observe the future behavior early, or set it to `WARN` to lock in the current behavior.
~~~

This deprecation log will be emitted when:

1) the user has explicitly configured `api.auth.type: basic`; AND
2) the user has _not_ specified `api.auth.basic.password_policy.mode`

We wish to change the _default_ value for `api.auth.basic.password_policy.mode` from its current `WARN` to `ERROR` at some undetermined point in the future (both options are currently supported). By specifying either value explicitly, the user can lock in their desired behavior and will be no longer at risk of the default changing out of under them, and will therefore not receive any more deprecation noise.

**Review by yaauie — APPROVED** [review]
From a code standpoint, LGTM. This warns if and only if the user's configuration is in danger of breaking due to a planned change in default value for `api.auth.basic.password_policy.mode`.

I'm pinging @karenzone for a review of the wording.

File: logstash-core/lib/logstash/webserver.rb (hunk 2 above)
Comment (karenzone) [reply logstash-core/lib/logstash/webserver.rb:112]: A nice future-proof message. Thanks for the ping.

**Comment by mashhurs** [comment]
@logstashmachine backport 8.3

**Pull request merged**